network/src/peer_manager/mod.rs

```diff
@@ -459,50 +459,63 @@ where
-                // TODO: Allow for trusted peers to still connect
-                if conn.metadata.origin == ConnectionOrigin::Outbound
-                    || self
-                        .active_peers
-                        .contains_key(&conn.metadata.remote_peer_id)
-                    || inbound_conns < self.inbound_connection_limit
-                {
-                    info!(
-                        NetworkSchema::new(&self.network_context)
-                            .connection_metadata_with_address(&conn.metadata),
-                        "{} New connection established: {}", self.network_context, conn.metadata
-                    );
-                    // Add new peer, updating counters and all
-                    self.add_peer(conn);
-                    self.update_connected_peers_metrics();
-                } else {
-                    info!(
-                        NetworkSchema::new(&self.network_context)
-                            .connection_metadata_with_address(&conn.metadata),
-                        "{} Connection rejected due to connection limit: {}",
-                        self.network_context,
-                        conn.metadata
-                    );
-                    self.disconnect(conn);
+                let trusted_peers = self.trusted_peers.read().clone();
+
+                match conn.metadata.origin {
+                    ConnectionOrigin::Outbound => {
+                        // TODO: This is right now a hack around having to feed trusted peers deeper in the outbound path.  Inbound ones are assigned at Noise handshake time.
+                        conn.metadata.role = trusted_peers
+                            .get(&conn.metadata.remote_peer_id)
+                            .map_or(PeerRole::Unknown, |auth_context| auth_context.role);
+                    }
+                    ConnectionOrigin::Inbound => {
+                        // Everything below here is meant for unknown peers only, role comes from
+                        // Noise handshake and if it's not `Unknown` it is trusted
+                        if conn.metadata.role != PeerRole::Unknown {
+                            return;
+                        }
+                        // TODO: Keep track of somewhere else to not take this hit in case of DDoS
+                        // Count unknown inbound connections
+                        let unknown_inbound_conns = self
+                            .active_peers
+                            .iter()
+                            .filter(|(peer_id, (metadata, _))| {
+                                metadata.origin == ConnectionOrigin::Inbound
+                                    && trusted_peers
+                                        .get(peer_id)
+                                        .map_or(true, |peer| peer.role == PeerRole::Unknown)
+                            })
+                            .count();
```

Comment:
maybe pull this out into its own fn so we can easily swap in a field later. think it also makes this fn a bit easier to follow